up-l1/mqtt.adoc

```diff
@@ -35,8 +35,85 @@ MQTT is an OASIS standard messaging protocol for the Internet of Things (IoT). I
 
 For more information, please refer to https://mqtt.org/
 
-
 == Specifications
 
+=== UAttribute Mapping
+
+MQTT 5 supports custom header fields, and we leverage this to map certain UAttribute values into the MQTT header.
+
+[cols="1,1"]
+|===
+| MQTT Header Property Code | Value (key:value)
+| UserProperty
+| ("upriority","{UAttribute.priority}")
+| CorrelationData
+| ("reqId", "{UAttribute.reqId}")
+| UserProperty
+| ("sink","{UAttribute.sink}") # Should this be serialized to MicroUri?
+|===
+
+The upriority field is needed as mqtt's QoS is very limited, so this will allow listeners to prioritize messages.
+
+The Correlation Data Property Code contains the request correlation id.
+
+The sink field is used to correctly map a message to the correct uE as the topic name may not contain all of the information.
+
+=== URI Mapping
+
+Depending on whether the UURI is local or remote, the MQTT Topic that the message will be published on will be different:
+
+* The remote UUri (with UAuthority): `upr/UUri.UAthority.name`
+* The local UUri (without UAuthority): `upl/UUri.UEntity/UUri.UResource`
+
+Examples:
+
+* The remote UUri (with UAuthority):
+
+[source]
+----
+UUri {
+  authority: UAuthority {
+    name: "Vehicle_1",
+    number: Id({01, 02, 03, 10, 11, 12})
+  },
+  entity: UEntity {
+    name: "body.access",
+    version_major: 1,
+    id: 1234,
+  }
+  resource: UResource {
+    name: "door",
+    instance: "front_left",
+    message: "Door",
+    id: 5678,
+  }
+}
+----
+
+Since this is a remote UUri, the MQTT Topic will simply be `upr/Vehicle_1`.
+
+* The local UUri (without UAuthority):
+
+[source]
+----
+UUri {
+  authority: empty,
+  entity: UEntity {
+    name: "body.access",
+    version_major: 1,
+    id: 1234,
+  }
+  resource: UResource {
+    name: "door",
+    instance: "front_left",
+    message: "Door",
+    id: 5678,
+  }
+}
+----
+
+Since this is a local UUri, the MQTT Topic will be constructed from the UEnity name and the UResource name `upl/body.access/door`.
```

Comment:
I'm wondering if it would be worth it to carry along the version_major inside the topic. That way, we could differentiate between people publishing to one version of the UEntity vs another. It could look something like this:

upl/body.access/1/door

Reply:
We need to use names.

Reply:
Does the UResource include the character '#'? In MQTT topics, '#' is considered a reserved character. The structure of the topic is crucial for MQTT client subscriptions and for performing authorization checks.  

We have to have different topic strategies for message type:

-  For pub message, the MQTT Topic will be constructed from the source uEntity name and the UResource 

-  For req, res, and notification message,  the MQTT Topic will be constructed from the sink uEntity name and the UResource

Including a version in the topic allows for better management of message routing and ensures compatibility between different versions of the services or payloads being communicated. This practice can be particularly useful in scenarios where you need to update the format or content of the messages being published without disrupting existing subscribers who might expect data in an older format.

Reply:
> Does the UResource include the character '#'? In MQTT topics, '#' is considered a reserved character. The structure of the topic is crucial for MQTT client subscriptions and for performing authorization checks.

[The long form does](https://github.com/eclipse-uprotocol/up-spec/blob/main/basics/uri.adoc#11-rfc3986) -- specifically for the "fragment", i.e. [the message](https://github.com/eclipse-uprotocol/up-core-api/blob/dc353878914b3459edf2e04531e9cadfc752a2e9/uprotocol/uri.proto#L96) portion of UResource.

Reply:
I have created a mapping of UUris to MQTT topic names in https://github.com/eclipse-uprotocol/up-spec/issues/73#issuecomment-1993445795.

Maybe that can serve as inspiration?

Reply:
@sophokles73 Overall I think that mapping would be good for cloud. Some questions would be:
1. Do we think the topic names would get to long in this scenario. If so, we could consider the short_form uri that @stevenhartley mentioned.
2. Would there be a reason to distinguish between local and remote uri's as currently proposed and implemented in Zenoh? I dont necessarily think there would beyond shortening the topic for local communication.

@PLeVasseur @lynn2009 Any thoughts? Ill go ahead and update the draft if not

Reply:
The mapping strategy is ok for the vehicle side. However, when considering the flow of messages between the device and the cloud, the mapping might cause confusion, even for the in-vehicle uStreamer. The in-vehicle uStreamer is configured to subscribe to any messages coming from the cloud to the vehicle using a wildcard, but it does not subscribe to messages it sends out. Due to the mapping, both C2D (Cloud-to-Device) and D2C (Device-to-Cloud) published message types begin with the same authority, leading to potential confusion.

To alleviate this issue, we could consider introducing a topic prefix, which seems like a viable option within UTransport. For instance, messages from the Cloud to Vehicle could use the prefix c2d/<authority>, while messages from the Vehicle to the Cloud could use d2c/<authority>.

Reply:
For topics that are frequently used, having a lengthy topic name can lead to increased bandwidth consumption during message publication. Utilizing MQTT 5.0 Topic Aliases could minimize that.  But there's typically a limitation on the number of aliases in a session.

Reply:
If I am understanding correctly, because a 'publish' message would use the device's uauthority it would cause confusion because any messages sent to the device from the cloud would start with the same uauthority. And then your suggestion would be to have a prefix like `c2d/` that the device would subscribe to with its uauthority id to get messages on? From the local message perspective, it may make sense to have some prefix like `upl` still.

Reply:
For the topic naming itself, I think if we serialize the UURI to short form we can just use that for the topic names (with the appropriate prefix). That should be short enough